File: engine/Utils/shopDataTypes.py
from enum import Enum
from .dataTypes import FrequencyCounter, WebsiteInfo


ITEMPROP = "itemprop"
ID = "id"
CLASS = "class"

TAG_ID_PRIORITY = [ITEMPROP, ID, CLASS]
REQUIRED_FIELDS = {"src", "width", "height", "id", "class", "itemprop", "data-a-dynamic-image"}
IDENTIFIERS_WITHOUT_CLASS = [ITEMPROP, ID]

SEPARATORS_SET = {"|", ":", "-"}

class PriceProperties(Enum):
    ID = 1
    TAG = 2
    PRICE = 3
    PRICE_ID = 4

class ImgProperties(Enum):
    ID = 1
    TAG = 2

class TitleProperties(Enum):
    ID = 1
    TAG = 2
    TITLE = 3
    FULL_TAG = 4

class ParsedProductProperties(Enum):
    TITLE = 1
    IMGS = 2
    PRICES = 3

class ProductProperties(Enum):
    TITLE = 1
    IMG = 2
    PRICE = 3

class ShopWebsiteInfo(WebsiteInfo):
    def __init__(self, url, name, url_id, hasJavascript, website_id=None, titleIds=None, imgIds=None):
        super().__init__(url, name, url_id, website_id)
        self.titleIdentifiers = titleIds if titleIds else set()
        self.imgIdentifiers = imgIds if imgIds else set()
        self.hasJavascript = hasJavascript

    def __str__(self):
        return "\n".join([
            f"{self.name}: {self.url}",
            f"  - Website ID: {self.website_id}",
            f"  - Favicon Name: {self.url_id}.ico",
            f"  - Has Javascript? {self.hasJavascript}",
            f"  - Titles IDs: {self.titleIdentifiers}",
            f"  - Images IDs: {self.imgIdentifiers}"
        ])

    def _addToIdentifiers(self, tagObjectfied, identifiersSet):
        addedAtLeastOne = False
        added = []

        for id in IDENTIFIERS_WITHOUT_CLASS:
            tagClass = tagObjectfied.get(id, None)
            if(tagClass and tagClass not in identifiersSet):
                identifiersSet.add(id)

                added.append(tagClass)
                addedAtLeastOne = True
        
        if(not addedAtLeastOne and CLASS in tagObjectfied):
            tagClass = tagObjectfied[CLASS]
            if(tagClass not in identifiersSet):
                identifiersSet.add(tagClass)
                added.append(tagClass)
        
        return added

    def addToTitleIdentifiers(self, tagObjectfied):
        return self._addToIdentifiers(tagObjectfied, self.titleIdentifiers)
    
    def addToImgIdentifiers(self, tagObjectfied):
        return self._addToIdentifiers(tagObjectfied["parent"] if "parent" in tagObjectfied else tagObjectfied, self.imgIdentifiers)
    
    def updateWithProduct(self, product):
        if(product):
            return self.addToTitleIdentifiers(product[ProductProperties.TITLE][TitleProperties.TAG]), self.addToImgIdentifiers(product[ProductProperties.IMG][ImgProperties.TAG])
    
    @classmethod
    def from_db(cls, website_id, url, name, url_id, hasJavascript, titleIds, imgIds):
        return cls(url, name, url_id, True if hasJavascript else False, website_id, titleIds, imgIds)
        
class ShopWebsitesInfo:
    def __init__(self):
        self._websiteInfo = {}
    
    def addWebsiteInfo(self, websiteInfo):
        self._websiteInfo[websiteInfo.url] = websiteInfo
    
    def removeWebsiteInfo(self, url):
        del self._websiteInfo[url]
    
    def get(self, url, notFound):
        return self._websiteInfo.get(url, notFound)

def getTagDictId(objectfied):
    if("tag" in objectfied):
        arr = [objectfied["tag"]]

        for id in TAG_ID_PRIORITY:
            if(id in objectfied): arr.append(f'{id}="{objectfied[id]}"')
        
        return " ".join(arr)
    else: return None

def get_product_obj(title, img, price):
    return {
        ProductProperties.TITLE: title,
        ProductProperties.IMG: img,
        ProductProperties.PRICE: price
    }

File: gui/linkers/modules/utils/shopDataTypes.js
import { WebsiteInfo } from "./dataTypes.js"
import { get_domain } from "../utils/websiteUtils.js"
import { get } from "./utils.js"

const ITEMPROP = "itemprop"
const ID = "id"
const CLASS = "class"

const TAG_ID_PRIORITY = [ITEMPROP, ID, CLASS]
const REQUIRED_FIELDS = new Set(["src", "width", "height", "id", "class", "itemprop", "data-a-dynamic-image", "data-src", "data-desktop-src"])
const IDENTIFIERS_WITHOUT_CLASS = [ITEMPROP, ID]

const SEPARATORS_SET = new Set(["|", "{", "-"])

const PriceProperties = {
    ID: 1,
    TAG: 2,
    PRICE: 3,
    PRICE_ID: 4
}

const ImgProperties = {
    ID: 1,
    TAG: 2
}

const TitleProperties = {
    ID: 1,
    TAG: 2,
    TITLE: 3,
    FULL_TAG: 4
}

const ParsedProductProperties = {
    TITLE: 1,
    IMGS: 2,
    PRICES: 3,
    BREADCRUMB: 4
}

const ProductProperties = {
    TITLE: 1,
    IMG: 2,
    PRICE: 3,
    URL: 4,
    BREADCRUMB: 5
}

class ShopWebsiteInfo extends WebsiteInfo{
    constructor(id, url, domain, name, hasJavascript=false, titleIds=null, imgIds=null){
        super(id, url, domain, name)
        this.titleIdentifiers = titleIds? titleIds: new Set()
        this.imgIdentifiers = imgIds? imgIds: new Set()
        this.hasJavascript = hasJavascript
    }

    toString(){
        return [
            `${this.name}: ${this.url}`,
            `  - Website ID: ${this.website_id}`,
            `  - Favicon Name: ${this.url_id}.ico`,
            `  - Has Javascript? ${this.hasJavascript}`,
            `  - Titles IDs: ${this.titleIdentifiers}`,
            `  - Images IDs: ${this.imgIdentifiers}`
        ].join("\n")
    }

    _addToIdentifiers(tagObjectfied, identifiersSet){
        let addedAtLeastOne = false
        let added = []
        let tag = tagObjectfied["tag"]

        for (let id of IDENTIFIERS_WITHOUT_CLASS){
            let tagClass = get(tagObjectfied, id, null)
            if(tagClass){
                tagClass = `${id}.${tagClass}`

                if(!identifiersSet.has(tagClass)){
                    identifiersSet.add(id)

                    added.push(tagClass)
                    addedAtLeastOne = true
                }else return []
            }
        }
        
        if(!addedAtLeastOne){
            let tagClass = tagObjectfied[CLASS]
            if(tagClass){
                tagClass = `${CLASS}.${tagClass}`
                
                if(!identifiersSet.has(tagClass)){
                    identifiersSet.add(tagClass)
                    added.push(tagClass)
                }
            }
        }
        
        return added
    }

    addToTitleIdentifiers(tagObjectfied){
        return this._addToIdentifiers(tagObjectfied, this.titleIdentifiers)
    }
    
    addToImgIdentifiers(tagObjectfied){
        return this._addToIdentifiers("parent" in tagObjectfied? tagObjectfied["parent"]: tagObjectfied, this.imgIdentifiers)
    }
    
    updateWithProduct(product){
        if(product)
            return [this.addToTitleIdentifiers(product[ProductProperties.TITLE][TitleProperties.TAG]), this.addToImgIdentifiers(product[ProductProperties.IMG][ImgProperties.TAG])]
    }
    
    static from_db(info, titleIds, imgIds){
        return new this(info.id, info.url, get_domain(info.url), info.name, info.hasJavascript? true:false, titleIds, imgIds)
    }
}
        
class ShopWebsitesInfo{
    _websiteInfo = {}

    addWebsiteInfo(websiteInfo){
        this._websiteInfo[websiteInfo.url] = websiteInfo
    }
    
    removeWebsiteInfo(url){
        delete this._websiteInfo[url]
    }
    
    get(url, notFound){
        return get(this._websiteInfo, url, notFound)
    }
}

function getTagDictId(objectfied){
    if("tag" in objectfied){
        let arr = [objectfied["tag"]]

        for (let id of TAG_ID_PRIORITY)
            if(id in objectfied) arr.push(`${id}="${objectfied[id]}"`)
        
        return arr.join(" ")
    }else return null
}    

function get_product_obj(title, img, price){
    return {
        [ProductProperties.TITLE]: title,
        [ProductProperties.IMG]: img,
        [ProductProperties.PRICE]: price
    }
}

export { ITEMPROP, ID, CLASS, TAG_ID_PRIORITY, REQUIRED_FIELDS, IDENTIFIERS_WITHOUT_CLASS, SEPARATORS_SET,
    PriceProperties, ImgProperties, TitleProperties, ParsedProductProperties, ProductProperties, ShopWebsiteInfo,
    ShopWebsitesInfo, getTagDictId, get_product_obj }